src/front/wgsl/mod.rs

```diff
@@ -328,241 +334,291 @@ impl<'a> Error<'a> {
                     (
-                        current.clone(),
-                        format!("redefinition of `{}`", &source[current.clone()]).into(),
+                        current,
+                        format!("redefinition of `{}`", &source[current]).into(),
                     ),
                     (
-                        previous.clone(),
-                        format!("previous definition of `{}`", &source[previous.clone()]).into(),
+                        previous,
+                        format!("previous definition of `{}`", &source[previous]).into(),
                     ),
                 ],
                 notes: vec![],
             },
+            Error::RecursiveDeclaration { ident, usage } => ParseError {
+                message: format!("declaration of `{}` is recursive", &source[ident]),
+                labels: vec![(ident, "".into()), (usage, "uses itself here".into())],
+                notes: vec![],
+            },
+            Error::CyclicDeclaration { ident, ref path } => ParseError {
+                message: format!("declaration of `{}` is cyclic", &source[ident]),
+                labels: path
+                    .iter()
+                    .enumerate()
+                    .flat_map(|(i, &(ident, usage))| {
+                        [
+                            (ident, "".into()),
+                            (
+                                usage,
+                                if i == path.len() - 1 {
+                                    "ending the cycle".into()
+                                } else {
+                                    format!("uses `{}`", &source[ident]).into()
+                                },
+                            ),
+                        ]
+                    })
+                    .collect(),
+                notes: vec![],
+            },
+            Error::ConstExprUnsupported(span) => ParseError {
+                message: "this constant expression is not supported".to_string(),
+                labels: vec![(span, "expression is not supported".into())],
+                notes: vec!["this should be fixed in a future version of Naga".into()],
+            },
+            Error::InvalidSwitchValue { uint, span } => ParseError {
+                message: "invalid switch value".to_string(),
+                labels: vec![(
+                    span,
+                    if uint {
+                        "expected unsigned integer"
+                    } else {
+                        "expected signed integer"
+                    }
+                    .into(),
+                )],
+                notes: vec![if uint {
+                    format!("suffix the integer with a `u`: '{}u'", &source[span])
+                } else {
+                    let span = span.to_range().unwrap();
+                    format!(
+                        "remove the `u` suffix: '{}'",
+                        &source[span.start..span.end - 1]
```

Comment:
There's no way this could end up falling in the midst of a multibyte UTF-8 character, right?

Reply:
Yeah the last character is always guaranteed to be a `u`.